---
title: "2-DESeq2 analysis"
date: "`r format (Sys.time(), format='%Y-%m-%d %H:%M:%S %z')`"
bibliography: bib.bib
output: 
  #bookdown::pdf_book:
  pdf_document:
    # base_format: rticles::springer_article
    pandoc_args:
      - '../../common.yaml'
      - --listings
    includes: 
      in_header: '../../preamble.tex'
    toc: true
    number_sections: true
    fig_width: 7
    fig_height: 6
    fig_caption: true
    df_print: tibble
---

```{r include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

\newpage

# Description
RNA-seq data were analyzed using R Bioconductor (3.5.1) and DESeq2 package (version 1.26.0)[@DESeq2]. 

# Load packages and data

```{r}
library(DESeq2)
library(ggplot2)
library(pheatmap)
library(RColorBrewer)
library(EnhancedVolcano)
library(forcats)
```

Counts data are also accessible in NCBI GEO under accession number GSE183973. 

```{r}
COUNTS <- read.table("./merged_gene_counts.txt",sep="\t", header=T, row.names = NULL)

dim(COUNTS)
```

Make gene names as rownames: 
```{r}
Genes <- COUNTS$gene_name
rownames(COUNTS) = make.names(Genes, unique=TRUE)

COUNTS <- COUNTS[,-c(1:2)]
head(COUNTS, 3)

```

Arrange the sample order to have the right group order: Healthy, Smoker and COPD. 

```{r}
COUNTS <- COUNTS [,c(4,1,15,7,12,21,26,17,27,3,13,24,6,23,18,10,19,20,2,22,9,8,14,5,16,25,11)]
```

# Make metadata for bulkRNAseq samples 

```{r}
colnames(COUNTS) <- c("Healthy_1_Mono", "Healthy_1_cAM", "Healthy_1_sAM", "Healthy_2_Mono", "Healthy_2_cAM", "Healthy_2_sAM", "Healthy_3_Mono", "Helathy_3_cAM", "Healthy_3_sAM", "Smoker_1_Mono", "Smoker_1_cAM", "Smoker_1_sAM", "Smoker_2_Mono", "Smoker_2_cAM", "Smoker_2_sAM", "Smoker_3_Mono", "Smoker_3_cAM", "Smoker_3_sAM", "COPD_1_Mono", "COPD_1_cAM", "COPD_1_sAM","COPD_2_Mono", "COPD_2_cAM", "COPD_2_sAM", "COPD_3_Mono","COPD_3_cAM", "COPD_3_sAM")

SampleSheet <- data.frame(
  "Treatment" = rep(c("Healthy","Smoker","COPD"),each=9),
  
  "Cells" = rep(c("Monocytes","AFhi cAM","AFlo AM"),3)
)

SampleSheet
```

```{r}
rownames(SampleSheet) <- colnames(COUNTS)
SampleSheet
```

# DESeq2

```{r}
dds <- DESeqDataSetFromMatrix(
  countData= COUNTS,
  colData= SampleSheet,
  design= ~ Cells + Treatment
)

dds
```

## Perform rlog transformation for distances and PCA

```{r}
# keep only genes with more than a single read
dds <- dds[ rowSums(counts(dds)) > 1,]

# perform rlog transformation for distances (for clustering) and PCA
rld<-rlog(dds)
```

```{r}
dds <- dds[ rowSums(counts(dds)) > 1,]
nrow(dds)
```


Calculate sample-to-sammple distances

```{r}
sampleDists <- dist( t( assay(rld) ) )
sampleDistMatrix <- as.matrix( sampleDists ) 
```

## Heatmap

```{r fig.width=8, fig.height=8}
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
heatmap <- pheatmap(sampleDistMatrix,
                    clustering_distance_rows=sampleDists,
                    clustering_distance_cols=sampleDists,
                    col=colors
)
```

## PCA analysis

```{r}
plotPCA <- plotPCA(rld, intgroup = c("Cells","Treatment"))
plotPCA
```

## Differentially expressed (DE) genes in comparing AFlo vs AFhi alveolar macrophages

```{r}
dds1 <- DESeq(dds)
res_AFlo_vs_AFhi<- results(dds1, contrast=c("Cells","AFlo AM","AFhi cAM"), lfcThreshold = 1, alpha = 0.05)
summary(res_AFlo_vs_AFhi) 
```

```{r}
Res_AFlo_vs_AFhi_Shrunk <- lfcShrink(dds1, contrast=c("Cells","AFlo AM","AFhi cAM"), res=res_AFlo_vs_AFhi, type = "normal")

AFlo_vs_AFhi <- merge(x=as.data.frame(res_AFlo_vs_AFhi), y=as.data.frame(Res_AFlo_vs_AFhi_Shrunk), by=c(0,1))

head(AFlo_vs_AFhi)
```



# Export DE genes for other analyses

```{r}
Genes2 <- AFlo_vs_AFhi$Row.names
head(Genes2, 3)
```

```{r}
rownames(AFlo_vs_AFhi) = make.names(Genes2, unique=TRUE)
AFlo_vs_AFhi<- AFlo_vs_AFhi[,-1]
```


Filter
```{r}
AFlo_vs_AFhi <- AFlo_vs_AFhi[!is.na(AFlo_vs_AFhi$padj.y),]
AFlo_vs_AFhi_1 <- subset(AFlo_vs_AFhi, padj.y < 0.05)
dim(AFlo_vs_AFhi_1)
```

```{r}
AFlo_vs_AFhi_ordered <- AFlo_vs_AFhi_1[order(-AFlo_vs_AFhi_1$log2FoldChange.y) , ]
AFlo_vs_AFhi_ordered
```

Save data for other analyses

```{r eval=FALSE}
write.table(as.data.frame(AFlo_vs_AFhi_ordered), "Results_Mreg_MA_LFC_9patients.txt", sep="\t", row.names=T,col.names=T)

```

# Volcano plots

```{r}
keyvals <- rep("black", nrow(AFlo_vs_AFhi))
names(keyvals) <- rep("non-signif", nrow(AFlo_vs_AFhi))

keyvals[which(AFlo_vs_AFhi$log2FoldChange.y > 1 )] <- "#ff8e03"
names(keyvals)[which(AFlo_vs_AFhi$log2FoldChange.y > 1)] <- "AFlo AM"

keyvals[which(AFlo_vs_AFhi$log2FoldChange.y < -1)] <- '#371dad'
names(keyvals)[which(AFlo_vs_AFhi$log2FoldChange.y < -1)] <- "AFhi cAM"

```

```{r fig.width=6, fig.height=5}
EnhancedVolcano(AFlo_vs_AFhi,
                lab = rownames(AFlo_vs_AFhi),
                x = 'log2FoldChange.y',
                y = 'padj.y',
                xlim = c(-15, 15),
                ylim=c(0, -log10(10e-120)),
                labSize = 0,
                pCutoff = 0.05,
                FCcutoff = 1,
                colAlpha = 1,
                colCustom = keyvals,
                legendLabSize = 8,
                legendIconSize = 2.0,
                border = "full",
                legendPosition = "right",
                axisLabSize = 20)
```

## Plot Macrophages- and Monocytes-associated genes

```{r fig.width=5, fig.height=5}
#plotCount MARCO
data <- plotCounts(dds, gene="MARCO", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("MARCO") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count")+
  theme_linedraw()+
  theme_light()

```


```{r fig.width=5, fig.height=5}
#plotCount FABP4
data <- plotCounts(dds, gene="FABP4", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("FABP4") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count")+
  theme_bw()+
  theme_linedraw()+
  theme_light()
```

```{r fig.width=5, fig.height=5}
#plotCount PPARG
data <- plotCounts(dds, gene="PPARG", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("PPARG") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count")+
  theme_bw()+
  theme_linedraw()+
  theme_light()
```

```{r fig.width=5, fig.height=5}
#plotCount MRC1
data <- plotCounts(dds, gene="MRC1", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("MRC1") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count")+
  theme_bw()+
  theme_linedraw()+
  theme_light()
```


```{r fig.width=5, fig.height=5}
#plotCount CCR2
data <- plotCounts(dds, gene="CCR2", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("CCR2") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count") + 
  theme_bw() +
  theme_linedraw()+
  theme_light()
```

```{r fig.width=5, fig.height=5}
#plotCount CX3CR1
data <- plotCounts(dds, gene="CX3CR1", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("CX3CR1") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count")+ 
  theme_bw()+
  theme_linedraw()+
  theme_light()
```


```{r fig.width=5, fig.height=5}
#plotCount ADAM19
data <- plotCounts(dds, gene="ADAM19", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("ADAM19") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count")+
  theme_bw()+
  theme_linedraw()+
  theme_light()
```

```{r fig.width=5, fig.height=5}

#plotCount CD14
data <- plotCounts(dds, gene="CD14", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("CD14") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count")+
  theme_bw() +
  theme_linedraw()+
  theme_light()
```
```{r fig.width=5, fig.height=5}
#plotCount IL10
data <- plotCounts(dds, gene="IL10", intgroup=c("Treatment","Cells"), returnData=TRUE)
ggplot(data, aes(x=Cells, y=count, color=Cells, shape = Treatment)) +
  scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0), size =3) +
  ggtitle("IL10") +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(values = c("#371dad","#ff8e03","#74c72a"))+
  ylab ("Normalized Count")+ 
  theme_bw()+
  theme_linedraw()+
  theme_light()
```

# DESeq2 analysis for AF^lo^ AM vs Monocytes

```{r}
res_AFlo_vs_Monocytes<- results(dds1, contrast=c("Cells","AFlo AM","Monocytes"), lfcThreshold = 1, alpha = 0.05)


res_AFlo_vs_Monocytes
summary(res_AFlo_vs_Monocytes) 

```

```{r}
#Shrunk
Res_AFlo_vs_Monocytes_Shrunk <- lfcShrink(dds1, contrast=c("Cells","AFlo AM","Monocytes"),type = "normal",  res=res_AFlo_vs_Monocytes)
```

```{r}
#ajouter colonne

AFlo_vs_Monocytes <- merge(x=as.data.frame(res_AFlo_vs_Monocytes), y=as.data.frame(Res_AFlo_vs_Monocytes_Shrunk), by=c(0,1))

#changer nom des colonne
Genes2 <- AFlo_vs_Monocytes$Row.names
rownames(AFlo_vs_Monocytes) = make.names(Genes2, unique=TRUE)
```

```{r}
AFlo_vs_Monocytes<- AFlo_vs_Monocytes[,-1]
```


```{r}
# filter 
# remove les pvalue NA
AFlo_vs_Monocytes <- AFlo_vs_Monocytes[!is.na(AFlo_vs_Monocytes$padj.y),] # 20580

#Enlever les pvalue >0.05
AFlo_vs_Monocytes_1 <- subset(AFlo_vs_Monocytes, padj.y < 0.05)
```

```{r eval=FALSE}
#To save
write.table(as.data.frame(AFlo_vs_Monocytes_ordered), "Results_Mreg_Monocytes_LFC_9patients.txt", sep="\t", row.names=T,col.names=T)

```


## Volcano plots for comparaison of AF^lo^ AM vs Monocytes


```{r}
keyvals <- rep("black", nrow(AFlo_vs_Monocytes))
names(keyvals) <- rep("non-signif", nrow(AFlo_vs_Monocytes))

keyvals[which(AFlo_vs_Monocytes$log2FoldChange.y > 1 )] <- "#ff8e03"
names(keyvals)[which(AFlo_vs_Monocytes$log2FoldChange.y > 1)] <- "AFlo AM"

keyvals[which(AFlo_vs_Monocytes$log2FoldChange.y < -1)] <- '#74c72a'
names(keyvals)[which(AFlo_vs_Monocytes$log2FoldChange.y < -1)] <- "Monocytes"

```

```{r fig.width=6, fig.height=5}
EnhancedVolcano(AFlo_vs_Monocytes,
                lab = rownames(AFlo_vs_Monocytes),
                x = 'log2FoldChange.y',
                y = 'padj.y',
                xlim = c(-15, 15),
                ylim=c(0, -log10(10e-300)),
                labSize = 0,
                pCutoff = 0.05,
                FCcutoff = 1,
                colAlpha = 1,
                colCustom = keyvals,
                legendLabSize = 8,
                legendIconSize = 2.0,
                border = "full",
                legendPosition = "right",
                axisLabSize = 20)

```


# DESeq2 analysis for comparaison of Monocytes vs AF^hi^ AM

```{r}
res_Monocytes_vs_AFhi<- results(dds1, contrast=c("Cells","Monocytes","AFhi cAM"), lfcThreshold = 1, alpha = 0.05)
#Shrunk
Res_Monocytes_vs_AFhi_Shrunk <- lfcShrink(dds1, contrast=c("Cells","Monocytes","AFhi cAM"), type = "normal", res=res_Monocytes_vs_AFhi)

#ajouter colonne
Monocytes_vs_AFhi <- merge(x=as.data.frame(res_Monocytes_vs_AFhi), y = as.data.frame(Res_Monocytes_vs_AFhi_Shrunk), by=c(0,1))

#changer nom des colonne
Genes2 <- Monocytes_vs_AFhi$Row.names
rownames(Monocytes_vs_AFhi) = make.names(Genes2, unique=TRUE)
Monocytes_vs_AFhi<- Monocytes_vs_AFhi[,-1]

# filter 
# remove les pvalue NA
Monocytes_vs_AFhi <- Monocytes_vs_AFhi[!is.na(Monocytes_vs_AFhi$padj.y),]

#Enlever les pvalue >0.05
Monocytes_vs_AFhi_1 <- subset(Monocytes_vs_AFhi, padj.y < 0.05)

```

```{r eval=FALSE}
#To save
write.table(as.data.frame(Monocytes_vs_AFhi_ordered), "Results_Monocytes_MA_LFC_9patients.txt", sep="\t", row.names=T,col.names=T)
```


## Volcano plots for comparaison of Monocytes vs AF^hi^ AM

```{r}
keyvals <- rep("black", nrow(Monocytes_vs_AFhi))
names(keyvals) <- rep("non-signif", nrow(Monocytes_vs_AFhi))

keyvals[which(Monocytes_vs_AFhi$log2FoldChange.y > 1 )] <- "#74c72a"
names(keyvals)[which(Monocytes_vs_AFhi$log2FoldChange.y > 1)] <- "Monocytes"

keyvals[which(Monocytes_vs_AFhi$log2FoldChange.y < -1)] <- '#371dad'
names(keyvals)[which(Monocytes_vs_AFhi$log2FoldChange.y < -1)] <- "AFhi cAM"

```


```{r fig.width=6, fig.height=5}
EnhancedVolcano(Monocytes_vs_AFhi,
                lab = rownames(Monocytes_vs_AFhi),
                x = 'log2FoldChange.y',
                y = 'padj.y',
                xlim = c(-15, 15),
                ylim=c(0, -log10(10e-300)),
                labSize = 0,
                pCutoff = 0.05,
                FCcutoff = 1,
                colAlpha = 1,
                colCustom = keyvals,
                legendLabSize = 8,
                legendIconSize = 2.0,
                border = "full",
                legendPosition = "right",
                axisLabSize = 20)
```

# Session information

```{r}
sessionInfo()
```

# References
